File: Pandacap/CompositeNotificationHandler.cs
using Pandacap.Extensions;
using Pandacap.Notifications.Interfaces;
using Pandacap.UI.Badges;

namespace Pandacap
{
    public class CompositeNotificationHandler(
        IEnumerable<INotificationHandler> notificationHandlers
    )
    {
        public IAsyncEnumerable<INotification> GetNotificationsAsync() =>
            notificationHandlers
            .Select(handler => new NotificationFailureHandler(handler))
            .Select(handler => handler.GetNotificationsAsync())
            .MergeNewest(post => post.Timestamp);

        private class NotificationFailureHandler(INotificationHandler underlyingHandler)
        {
            public async IAsyncEnumerable<INotification> GetNotificationsAsync()
            {
                var enumerable = underlyingHandler.GetNotificationsAsync();
                var enumerator = enumerable.GetAsyncEnumerator();

                INotification? initial = null;
                INotification? error = null;

                try
                {
                    if (await enumerator.MoveNextAsync())
                        initial = enumerator.Current;
                }
                catch (Exception ex)
                {
                    error = new ErrorNotification(ex, underlyingHandler);
                }

                if (error != null)
                {
                    yield return error;
                    yield break;
                }

                if (initial != null)
                    yield return initial;

                while (await enumerator.MoveNextAsync())
                    yield return enumerator.Current;
            }
        }

        private record ErrorNotification(
            Exception Exception,
            INotificationHandler UnderlyingHandler) : INotification
        {
            public string? ActivityName => Exception.GetType().Name;
            public Badge Badge => Badges.Pandacap;
            public string? Url => null;
            public string? UserName => $"{UnderlyingHandler.GetType().Name}: {Exception.Message}";
            public string? UserUrl => null;
            public string? PostUrl => null;
            public DateTimeOffset Timestamp => DateTime.UtcNow;
        }
    }
}

File: Pandacap.Notifications.Composite/CompositeNotificationHandler.fs
namespace Pandacap.Notifications.Composite

open System
open FSharp.Control
open Pandacap.Extensions
open Pandacap.UI.Badges
open Pandacap.Notifications.Composite.Interfaces
open Pandacap.Notifications.Interfaces

type internal CompositeNotificationHandler(
    notificationHandlers: INotificationHandler seq
) =
    let wrapInErrorHandler (underlyingHandler: INotificationHandler) = {
        new INotificationHandler with
            member _.GetNotificationsAsync() = asyncSeq {
                let enumerable = underlyingHandler.GetNotificationsAsync()
                let enumerator = enumerable.GetAsyncEnumerator()

                let mutable more = true
                while more do
                    more <- false

                    try
                        let! hasItem = enumerator.MoveNextAsync().AsTask() |> Async.AwaitTask
                        if hasItem then
                            yield enumerator.Current
                            more <- true
                    with ex ->
                        yield {
                            new INotification with
                                member _.ActivityName = ex.GetType().Name
                                member _.Badge = Badges.Pandacap
                                member _.Url = null
                                member _.UserName = $"{underlyingHandler.GetType().Name}: {ex.Message}"
                                member _.UserUrl = null
                                member _.PostUrl = null
                                member _.Timestamp = DateTimeOffset.UtcNow
                        }
        }
    }

    interface ICompositeNotificationHandler with
        member _.GetNotificationsAsync() =
            let sequences =
                notificationHandlers
                |> Seq.map wrapInErrorHandler
                |> Seq.map (fun handler -> handler.GetNotificationsAsync())

            sequences.MergeNewest(fun notification -> notification.Timestamp)
